From 54df04f09cb084b9e58529c0ae6f53f0e50f1a19 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Domen=20Ko=C5=BEar?= <domen@cachix.org>
Date: Fri, 11 Jul 2025 11:34:59 -0500
Subject: [PATCH 1/2] Always allow 'builtins.getEnv SECRETSPEC_SECRETS'.

This way we avoid writing them to a file, but they still leak to drvs.
---
 src/libexpr/primops.cc | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/src/libexpr/primops.cc b/src/libexpr/primops.cc
index 6b8b96d4c..faed8be70 100644
--- a/src/libexpr/primops.cc
+++ b/src/libexpr/primops.cc
@@ -1037,7 +1037,13 @@ static void prim_getEnv(EvalState & state, const PosIdx pos, Value * * args, Val
 {
     std::string name(state.forceStringNoCtx(*args[0], pos, "while evaluating the first argument passed to builtins.getEnv"));
     printTalkative("devenv getEnv: '%1%'", name);
-    v.mkString(state.settings.restrictEval || state.settings.pureEval ? "" : getEnv(name).value_or(""));
+
+    // Allow SECRETSPEC_SECRETS even in pure/restricted mode
+    if (name == "SECRETSPEC_SECRETS" || !(state.settings.restrictEval || state.settings.pureEval)) {
+        v.mkString(getEnv(name).value_or(""));
+    } else {
+        v.mkString("");
+    }
 }
 
 static RegisterPrimOp primop_getEnv({
-- 
2.49.0

